docs/en/creating-widgets/supplemental.md

```diff
@@ -0,0 +1,869 @@
+-   Internal or private state can be fully encapsulated within a widget, if required.
+    -   It is a valid and often desireable pattern to implement 'pure' widgets that incur no side effects and receive their entire state as properties, but this is not the only pattern in Dojo widget development.
+
+## Using class-based widgets
+
+-   The _`__render__`_, _`__setProperties__`_, and _`__setChildren__`_ functions are internal framework implementation details and should never be called nor overridden in application code.
+-   Applications should not instantiate widgets directly - Dojo fully manages the lifecycle of widget instances, including instantiation, caching and destruction.
+    -   Applications only need to provide widgets to the framework as generic type parameters when calling [the `w()` method](#instantiating-vdom-nodes), or implicitly via [TSX tags](#tsx-widget-example), as VDOM output from widgets' render functions.
+
+## The Virtual DOM
+
+-   Virtual node [`key`s](#vdom-node-keys) should be consistent across multiple render calls.
+    -   If a different `key` is specified on every render invocation, Dojo cannot efficiently associate nodes across previous and current renders. Dojo treats new `key`s that it has not seen in the previous render as new elements, which results in the previous nodes being removed from the DOM and an entirely new set being added - even when there are no other property changes that would actually warrant a DOM update.
+    -   A common anti-pattern is assigning a randomly-generated ID (such as a GUID or UUID) as a node's `key` within a widget's render function. Node `key`s should not be generated within a render function unless the generation strategy is idempotent.
+-   Applications should not store references to virtual nodes for later manipulation outside of returning them from a widget's render function, nor as an attempt to optimize memory allocation by using a single instance across multiple render calls.
+    -   Virtual nodes are designed to be lightweight and cost very little to instantiate new versions of within every widget render cycle.
+    -   The framework relies on having two separate virtual node instances across two widget render function calls to perform accurate change detection. If no changes are detected, no further cost is incurred, rendering or otherwise.
+
+## Rendering and the DOM
+
+-   Applications should have no need to use imperative DOM manipulation calls.
+    -   The framework handles all concrete rendering responsibilities, and provides alternative mechanisms to widget authors for using a variety of DOM functionality in a more simplified, type-safe, and reactive manner.
+-   Dojo does not support changing event listener functions after the first render.
+    -   Only changes to node properties and attributes are reflected against DOM nodes between subsequent render cycles - listener functions that need to vary their implementation between renders should be avoided.
```

Comment:
"listener functions that need to vary their implementation between renders should be avoided." I don't think that this is true, a listener can be changed every render.

Reply:
👍 more outdated info from the existing docs. Removed.